tests/thread_local/thread_local.edl

```diff
@@ -0,0 +1,16 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+enclave {
+    trusted {
+        public void enclave_thread(
+            int thread_num, 
+            int iters, 
+            int step, 
+            [out] int** tls_value1_addr, 
```

Comment:
I have seen this in a series of other recent tests as well. What is the reasoning behind using [out] int** instead of [out] int*? EDL will take care of the transition between host and enclave memory. [out] int** forces the enclave to allocate host memory and leads to pointer management problems (since the inner pointer is not validated by EDL). It just seems like a bad pattern to showcase to novice users of the SDK.

Reply:
`[user_check, in] type*` is acceptable for asynchronous communication between the host and the enclave, but the enclave should then **check** the pointer (as the specification provided in the EDL states).

`[out, size=…] type*` is acceptable to return strings or other arrays outside from the enclave.

The only valid use case for `[out] type**` would be multi-dimensional data, which is best flattened by the user, since EDL does not support nested pointers inherently.

I know that this is test code and that security does no matter, but users should be able to look at the tests as clean and well written examples of OpenEnclave usage.

Reply:
@ihaller In this test, the enclave is returning the **addresses** of couple of thread-local variables of each enclave thread back to the host so that the host can assert that the addresses of the thread-local variables are different across threads.

Reply:
@ihaller Good feedback. I have rewritten the test to avoid passing addresses back.

Reply:
For that particular requirement the suggested pattern should be `[out] intptr_t*`. That clearly specifies that the returned value is a pointer-sized opaque handle where the only meaningful operation applied to it should be an equality comparison. Similarly, host handles passed into the enclave should be marked as `[in] intptr_t` (strings to be passed back to the host, socket/file identifiers, etc.).